### PR TITLE
Withdrawal tests

### PR DESCRIPTION
**Pull request by byteZorvin**
- This adds withdrawal limit component tests

### CODE REVIEW

File: src/withdrawal_limit/mock.cairo
Comment (apoorvsadana) [inline src/withdrawal_limit/mock.cairo:1]: should this move to a mock folder

File: src/withdrawal_limit/mock.cairo [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    #[constructor]
+    pub fn constructor(ref self: ContractState) {
+        self.withdrawal.initialize(5);
+    }
+
+
+    #[abi(embed_v0)]
+    impl MockWithdrawalLimitImpl of super::IMockWithdrawalLimit<ContractState> {
+        fn change_withdrawal_limit_token(
```
Comment (apoorvsadana) [inline src/withdrawal_limit/mock.cairo:53]:
```suggestion
        fn toggle_withdrawal_limit_for_token(
```

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    let withdrawal_limit_mock = IWithdrawalLimitDispatcher {
+        contract_address: withdrawal_limit_mock_address
+    };
+
+    let mut spy = snf::spy_events();
+    (withdrawal_limit_mock, spy)
+}
+
+
+pub fn deploy_erc20(name: ByteArray, symbol: ByteArray) -> ContractAddress {
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:37]: we already have this function in other test cases, can we not move to utils and reuse?

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+#[test]
+fn get_remaining_withdrawal_quota_ok() {
+    let (withdrawal_limit, _) = deploy_withdrawal_limit();
+
+    let usdc_address = deploy_erc20("USDC", "USDC");
+    let usdc = IERC20Dispatcher { contract_address: usdc_address };
+    // Mocking deposits with the contract
+    usdc.transfer(withdrawal_limit.contract_address, 1000_000);
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:65]:
```suggestion
    usdc.transfer(withdrawal_limit.contract_address, 1_000_000);
```

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    // Should return 0 when withdrawal limit not applied
+    assert(
+        withdrawal_limit.get_remaining_withdrawal_quota(usdc_address) == BoundedInt::max(),
+        'Quota should be 0'
+    );
+    withdrawal_limit_mock.change_withdrawal_limit_token(usdc_address, true);
+
+    // Should return the default 5% of the balance of contract when limit applied
+    assert(
+        withdrawal_limit.get_remaining_withdrawal_quota(usdc_address) == 5000_0,
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:80]:
```suggestion
        withdrawal_limit.get_remaining_withdrawal_quota(usdc_address) == 50_000,
```

you can actually just use 100 and 5, it's simpler imo :)

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    let usdc_address = deploy_erc20("USDC", "USDC");
+    let usdc = IERC20Dispatcher { contract_address: usdc_address };
+    // Mocking deposits with the contract
+    usdc.transfer(withdrawal_limit.contract_address, 1000_000);
+
+    let withdrawal_limit_mock = IMockWithdrawalLimitDispatcher {
+        contract_address: withdrawal_limit.contract_address
+    };
+
+    // Should return 0 when withdrawal limit not applied
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:71]:
```suggestion
    // Should return BoundedInt::max() when withdrawal limit not applied
```

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    assert(
+        withdrawal_limit.get_remaining_withdrawal_quota(usdc_address) == BoundedInt::max(),
+        'Quota should be 0'
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:74]:
```suggestion
        'Quota should be BoundedInt::max()'
```

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        contract_address: withdrawal_limit.contract_address
+    };
+
+    withdrawal_limit_mock.change_withdrawal_limit_token(usdc_address, true);
+
+    assert(
+        withdrawal_limit.get_remaining_withdrawal_quota(usdc_address) == 5000_0,
+        'Quota should not be 0'
+    );
+    withdrawal_limit_mock.consume_quota(usdc_address, 10000);
```
Comment (apoorvsadana) [inline src/withdrawal_limit/tests/withdrawal_limit_test.cairo:104]:
```suggestion
    withdrawal_limit_mock.consume_quota(usdc_address, 10_000);
```
I believe the normal `_` practice is every 3 digits from the right. or, we can skip `_` and use smaller numbers like 100 as mentioned above

File: src/withdrawal_limit/mock.cairo
Comment (byteZorvin) [reply src/withdrawal_limit/mock.cairo:1]: Moved

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo (hunk 2 above)
Comment (byteZorvin) [reply src/withdrawal_limit/tests/withdrawal_limit_test.cairo:37]: Did not find a way to import files from the tests/ folder into src/ so had to replicate

File: src/withdrawal_limit/mock.cairo
Comment (apoorvsadana) [reply src/withdrawal_limit/mock.cairo:1]: is it pushed?

File: src/withdrawal_limit/tests/withdrawal_limit_test.cairo (hunk 2 above)
Comment (apoorvsadana) [reply src/withdrawal_limit/tests/withdrawal_limit_test.cairo:37]: can you create an issue for this w/ an `enhancement` tag and add a todo on code as well

**Pull request merged**